simulator/src/main/java/com/github/benmanes/caffeine/cache/simulator/policy/greedy_dual/GDWheelPolicy.java

```diff
@@ -0,0 +1,235 @@
+        policyStats.recordOperation();
+        Node node = data.get(key);
+        int w = 0;
+        if (node == null) {
+            while (event.weight() + currentSize > maximumSize) {
+                int ch0 = wheels[0].getNextIndex();
+                boolean fullRound = CH[0] > ch0 || !wheels[0].queueNotEmpty(ch0);
+                CH[0] = ch0;
+                AccessEvent q = wheels[0].evict(CH[0]);
+                if (fullRound) {
+                    migration(1);
+                }
+                boolean admit = admittor.admit(event, q);
+                if (admit) {
+                    data.remove(q.key());
+                    currentSize -= q.weight();
+                    policyStats.recordEviction();
+                } else {
+                    wheels[0].add(CH[0], q);
+                    break;
+                }
+            }
+            policyStats.recordWeightedMiss(event.weight());
+            for (int i = 0; i < nW; i++) {
+                if (Math.round(event.missPenalty() / Math.pow(nQ, i)) > 0) {
+                    w = i;
+                }
+            }
+            int q = (int) (Math.round(event.missPenalty() / Math.pow(nQ, w)) + CH[w]) % nQ;
+            wheels[w].add(q, event);
+            currentSize += event.weight();
+            data.put(key, new Node(event, w, q));
+        } else {
+            policyStats.recordWeightedHit(event.weight());
+            wheels[node.wheel].remove(node.q, node.event);
+            for (int i = 0; i < nW; i++) {
+                if (Math.round(node.event.missPenalty() / Math.pow(nQ, i)) > 0) {
+                    w = i;
+                }
+            }
+            int q = (int) (Math.round(event.missPenalty() / Math.pow(nQ, w)) + CH[w]) % nQ;
+            node.updateLoc(w, q);
+            wheels[w].add(q, node.event);
+        }
+
+    }
+
+    private void migration(int idx) {
+        CH[idx] = (CH[idx] + 1) % nQ;
+        if (CH[idx] == 0 && idx + 1 < nW) {
+            migration(idx + 1);
+        }
+        while (wheels[idx].queueNotEmpty(CH[idx])) {
+            AccessEvent p = wheels[idx].evict(CH[idx]);
+            int cr = ((int) p.missPenalty()) % ((int) Math.pow(nQ, idx));
+            int q = (int) (Math.round(cr / Math.pow(nQ, idx - 1)) + CH[idx - 1]) % nQ;
+            wheels[idx - 1].add(q, p);
+            Node node = data.get(p.key());
+            node.updateLoc(idx - 1, q);
+        }
+    }
+
+    @Override
+    public PolicyStats stats() {
+        return policyStats;
+    }
+
+    static class GDWheelSettings extends BasicSettings {
+        public GDWheelSettings(Config config) {
+            super(config);
+        }
+
+        public int numberOfWheels() {
+            return config().getInt("gd-wheel.nw");
+        }
+
+        public int numberOfQueues() {
+            return config().getInt("gd-wheel.nq");
+        }
+    }
+
+    static class CostWheel {
+        List<PriorityQueue<AccessEvent>> wheel;
```

Comment:
You might want to use `TreeSet` instead, which is implemented as a red-black tree. That has a O(lg n) operation for `remove(Object)`, whereas `PriorityQueue` is a heap with O(n) time.

> Implementation note: this implementation provides O(log(n)) time for the enqueuing and dequeuing methods
offer, poll, remove(), and add; linear time for the remove(Object) and contains(Object) methods; and constant time for the retrieval methods peek, element, and size.

Reply:
PriorityQueue + HashMap all the way :)

Reply:
In this case, `remove(Object)` is called on every cache hit to reorder the entry. Since that is an O(n) operation, the performance degrades as the cache size increases.

`PriorityQueue` is optimal when used a a queue, where specific entries are not commonly searched for. For example Java's `ScheduledExecutorService` uses a min-heap and marks the entry as dead when cancelled, rather then removing it. In that case the cheaper algorithmic cost is beneficial, especially since a scheduler rarely has millions of events. Here the remove + add to resort the entry on every hit could lead to higher execution times and make GDWheel appear slower than it might be in practice.

Reply:
I'm always forgetting that `PriorityQueue` misses `removeAt(int)` method, so adding `HashMap` does not solve the issue. BinaryHeap with node removal is needed.

Tombstones would create an ABA issue, so it might be easier to implement or borrow a binary heap if priorityqueue with removals is needed.

> Here the remove + add to resort the entry on every hit could lead to higher execution times and make GDWheel appear slower than it might be in practice

The paper suggests that Queue is a linked list:

>Since each queue is implemented by a doubly linked list, removing or inserting a given node requires only constant time

In other words, the queue needs:
* Fast append to `head`
* Fast removal from `tail`
* Fast removal from the middle

It is funny how `LinkedHashSet` almost works here. The sad part is `LinkedHashSet` does not expose `reverseIterator` even though it maintains double-linked list.

Reply:
I'll have to reread the GDWheel paper, as it has been a few years. I believe Apache Commons' `ListOrderedMap` is that more flexible variant that would be simple to adapt here, based on your comment. In our [TimerWheel](https://github.com/ben-manes/caffeine/blob/master/caffeine/src/main/java/com/github/benmanes/caffeine/cache/TimerWheel.java) we use intrusive linked lists for expiration to maintain an O(1) priority queue. However that code is very optimized while this is simulator code for algorithm analysis. Ideally the simulator policies would be clean, readable reference implementations that are easy to adapt for others. They don't tackle messy topics like concurrency, though, and are strictly for analysis for exploring each algorithm's insights.

Reply:
`ListOrderedMap` uses `ArrayList` which implies `O(N)` removals.

What do you think of the following?
The below structure efficiently implements "add to head" (via `put(k, v)`), "remove from tail" (via `removeEldestEntry` trick) and "remove".

```java
public class DequeWithRemoval<K, V> extends LinkedHashMap<K, V> {
    private static final Object SENTINEL = new Object();
    private @Nullable V lastValue;
    private boolean removeLastRequested;

    public DequeWithRemoval(int initialCapacity, float loadFactor) {
        super(initialCapacity, loadFactor, /* accessOrder */ false);
    }

    public V removeLast() {
        if (isEmpty()) {
            throw new NoSuchElementException();
        }
        removeLastRequested = true;
        // put triggers removeEldestEntry which is how we retrieve the latest element
        //noinspection unchecked
        ((Map) this).put(SENTINEL, SENTINEL);
        V lastValue = this.lastValue;
        this.lastValue = null;
        // Remove sentinel
        values().iterator().remove();

        return lastValue;
    }

    @Override
    protected boolean removeEldestEntry(Map.Entry<K, V> eldest) {
        if (!removeLastRequested) {
            return false;
        }
        lastValue = eldest.getValue();
        return true;
    }
}
```

Reply:
oh man, I didn't realize it was done so poorly in Apache. I suppose it's my fault for never using it an assuming it was on par if I ever found a need.

That's neat. I don't mind implementing intrusive linked lists. If I was to write implement this policy, I would include the prev/next references on the `Node` and handle it directly. This is done in many of the other policies for LRU ordering. The code is very simple and straightforward.

Reply:
Oh, it was [LinkedMap](https://github.com/rkanter/collections-generic/blob/master/src/main/java/org/apache/commons/collections15/map/LinkedMap.java) that was like `LinkedHashMap`.

Reply:
I will change it to TreeSet, since according to the paper it should be O(log(n)) in the worst case and amortized is O(1), so O(n) for a single operation is not ideal for the implemantation.
Thanks for the info.

Reply:
@himelbrand , could you please clarify where do you get `according to the paper it should be O(log(n))` from?

https://conglongli.github.io/paper/gdwheel-eurosys2015.pdf

> 3.2.2 GD-Wheel Time Complexity Analysis
>Since each
>queue is implemented by a doubly linked list, removing
>or inserting a given node requires only constant time. An
>insertion of a given object requires one list insertion. A
>reuse of a given object requires one list removal and one list
>insertion. Thus handling either an insertion or a reuse of an
>object takes only O(1) time in the worst case

I do not see a priority queue here. However, I must admit it is the first time I see the paper, so I can be wrong.

Reply:
@vlsi, the are 2 published versions of this paper, but I think this appears in both of them.
specifically regarding the priority queue, it is definitely mentioned in the pseudo code of the algorithm and a few extra locations.

regarding the time complexity, if you keep going a bit further in 3.2.2 you'll see that O(1) is amortized. I might have been confused, as I now see that they mentioned O(n), but I might remember the log(n) from the other version when compared to the gree dual approach.

Reply:
PriorityQueue is mentioned in `Algorithm 1 The GreedyDual Algorithm under Cao et al.’s implementation`
However, point of adding `GD-Wheel` policy is to replace priority queues with multiple linked lists.

O(n) is related to `migration` which indeed scans the queue.

>paper: As a result, **objects with different priorities** are stored in **different queues**

So I'm inlined that the queues should be implemented as linked lists so the implementation matches the paper.

Reply:
When I implemented it I did it according to [pdf](https://www.cs.cmu.edu/~conglonl/gdwheel-ladis2013.pdf).
which is the Algorithm 1 The GreedyDual Algorithm under Cao et al.’s implementation.
the paper I gave the link and that you cited, was not the one I was reading when I implemented the algorithm.
This is the reason for the confusion.

It could be nice to implement the other version too, or to add the possibility to switch between the 2 variants in the future.

Reply:
Let me try to spend some time reading the two papers and seeing if I can optimize it while matching the hit rates. I was planning on refactoring the code a bit to pull apart the long `record` method, e.g. to match the other policies which use `onHit`, `onMiss`, `evict`, etc. I'd also like to extract paper comments to help guide the reader, e.g. as done in `GdfsPolicy`. This will all take a bit longer, but the code is a good baseline to refactor on top of.